helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableReplica.java

```diff
@@ -19,9 +19,119 @@
  * under the License.
  */
 
+import org.apache.helix.model.ResourceConfig;
+import org.apache.helix.model.StateModelDefinition;
+
+import java.io.IOException;
+import java.util.Map;
+
 /**
- * A placeholder before we have the implementation.
- *
  * This class represents a partition replication that needs to be allocated.
  */
-public class AssignableReplica { }
+public class AssignableReplica implements Comparable<AssignableReplica> {
+  private final String _partitionName;
+  private final String _resourceName;
+  private final String _resourceInstanceGroupTag;
+  private final int _resourceMaxPartitionsPerInstance;
+  private final Map<String, Integer> _capacityUsage;
+  // The priority of the replica's state
+  private final int _statePriority;
+  // The state of the replica
+  private final String _replicaState;
+
```

Comment:
Should we add a field where a priority among replicas could be defined? I see the state priority field, and I suspect this would be used to prioritize top states over all other states, but what about same-state replicas?

Reply:
If 2 replicas have the same state, and the same partition, they will be the same from any perspective. And if you think about it, they are identical replicas.
This shall be OK since as long as we have the right number of objects, the assignment result will be good. Even the algorithm shift 2 identical replica objects' allocation, the final mapping will still be the same.
So, till now, I don't see the necessity of adding a priority other than the state priority.

Reply:
I think I was thinking about priority among different partitions. I don't know how the assignment logic works (is it part of this PR or is it TBD in another PR?), but I agree with you if we assign replicas by partition. 

I was just bringing up the fact that we may need to account partition-level or resource-level priority at assign time.

Reply:
The rebalance logic will take care of the resource, partition, and also state priority.
The priority we recorded in the replica is mainly for the top state. Since that is really something special.
In fact, I tried to use a boolean isTopState. But then I find the number of priority already exists and it is the most direct information. The model class shall not do extra calculation.